bd-session-replay/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+  pub fn new(
+    target: Box<dyn Target + Send + Sync>,
+    runtime_loader: &Arc<ConfigLoader>,
+  ) -> (Self, Sender<()>) {
+    // Limit the buffer size to 1 to reduce the risk of putting too much pressure on the
+    // application's main thread when dequeuing the screenshot actions from the channel and
+    // passing them to the platform layer, which performs the actual screenshotting on the main
+    // thread.
+    let (take_screenshot_tx, take_screenshot_rx) = channel(1);
+
+    let mut is_periodic_reporting_enabled =
+      session_replay::PeriodicWireframesEnabledFlag::register(runtime_loader).unwrap();
+    let reporting_interval_rate = session_replay::ReportingIntervalFlag::register(runtime_loader)
+      .unwrap()
+      .read_mark_update();
+
+    let mut is_screenshotting_enabled_flag =
+      session_replay::ScreenshotsEnabledFlag::register(runtime_loader).unwrap();
+
+    (
+      Self {
+        target,
+        is_periodic_reporting_enabled: is_periodic_reporting_enabled.read_mark_update(),
+        reporting_interval_rate: reporting_interval_rate.unsigned_abs(),
+        reporting_interval: None,
+        is_screenshotting_enabled: is_screenshotting_enabled_flag.read_mark_update(),
+        take_screenshot_rx,
+        is_periodic_reporting_enabled_flag:
+          session_replay::PeriodicWireframesEnabledFlag::register(runtime_loader).unwrap(),
+        reporting_interval_flag: session_replay::ReportingIntervalFlag::register(runtime_loader)
+          .unwrap(),
+        is_screenshotting_enabled_flag,
+      },
+      take_screenshot_tx,
+    )
+  }
+
+  fn create_interval(interval: Duration, fire_immediately: bool) -> tokio::time::Interval {
+    let mut interval = if fire_immediately {
+      tokio::time::interval(interval)
+    } else {
+      tokio::time::interval_at(Instant::now() + interval, interval)
+    };
+
+    log::debug!(
+      "session replay recorder interval is {:?}",
+      interval.period()
+    );
+
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+    interval
+  }
+
+  pub async fn run(&mut self) {
+    let shutdown_trigger = ComponentShutdownTrigger::default();
+    self
+      .run_with_shutdown(shutdown_trigger.make_shutdown())
+      .await;
+  }
+
+  pub async fn run_with_shutdown(&mut self, mut shutdown: ComponentShutdown) {
+    if self.reporting_interval.is_none() {
+      self.reporting_interval = Some(Self::create_interval(self.reporting_interval_rate, true));
+    }
+
+    let local_shutdown = shutdown.cancelled();
+    tokio::pin!(local_shutdown);
+
+    loop {
+      tokio::select! {
+        () = async {
+          if let Some(reporting_interval) = &mut self.reporting_interval {
+            reporting_interval.tick().await;
+          }
+        }, if self.reporting_interval.is_some() && self.is_periodic_reporting_enabled => {
+          log::debug!("session replay recorder capturing wireframe");
+          self.target.capture_wireframe();
```

Comment:
This one is less of a risk but technically if we get backed up we could be issuing N of these. Per my other question not sure how the platform layer is going to work but worth a comment around this if it isn't an issue.